dll/kernel32.cpp

```diff
@@ -781,6 +921,39 @@ namespace kernel32 {
 		return cbMultiByte;
 	}
 
+	unsigned int WIN_FUNC MultiByteToWideChar(unsigned int codePage, unsigned int dwFlags, const char *lpMultiByteStr, int cbMultiByte, uint16_t *lpWideCharStr, int cchWideChar) {
+		DEBUG_LOG("MultiByteToWideChar(codePage=%u, dwFlags=%u, multiByte=%d, wideChar=%d)\n", codePage, dwFlags, cbMultiByte, cchWideChar);
+
+		// assert (dwFlags == 1); // MB_PRECOMPOSED
+		int i = 0;
+		if (lpWideCharStr == 0) { // return required buffer length
+			while (lpMultiByteStr[i] != 0) {
+				i++;
+			}
+		} else { // else copy source into destination
+			while (i < cchWideChar) {
+				lpWideCharStr[i] = lpMultiByteStr[i];
+				i++;
+			}
+			lpWideCharStr[cchWideChar] = 0; // NUL terminate
+		}
+		return i + 1;
+	}
+
+	unsigned int WIN_FUNC GetStringTypeW(unsigned int dwInfoType, const char *lpSrcStr, int cchSrc, uint16_t *lpCharType) {
+		DEBUG_LOG("GetStringTypeW (dwInfoType=%u, lpSrcStr=%p, cchSrc=%i, lpCharType=%p)\n", dwInfoType, lpSrcStr, cchSrc, lpCharType);
+
+		int strLen = cchSrc < 0 ? strlen(lpSrcStr) : cchSrc;
+		int i = 0;
+		while (i < strLen) {
+			lpCharType[i] = lpSrcStr[i];
+			i++;
+		}
+		lpCharType[strLen] = 0; // NUL terminate
```

Comment:
maybe something like
```suggestion
		while (i < strLen) {
			char c = lpSrcStr[i];
			bool upper = ('A' <= c && c <= 'Z');
			bool lower = ('a' <= c && c <= 'z');
			bool alpha = (lower || upper);
			bool digit = ('0' <= c && c <= '9');
			bool space = (c == ' ' || c == '\n' || c == '\t' || c == '\r' || c == '\f' || c == '\v');
			bool blank = (c == ' ' || c == '\t');
			bool hex = (digit || ('A' <= c && c <= 'F') || ('a' <= c && c <= 'f'));
			bool cntrl = ((0 <= c && c < 0x20) || c == 127);
			bool punct = (!cntrl && !digit && !alpha);
			lpCharType[i] = (upper ? 1 : 0) | (lower ? 2 : 0) | (digit ? 4 : 0) | (space ? 8 : 0) | (punct ? 0x10 : 0) | (cntrl ? 0x20 : 0) | (blank ? 0x40 : 0) | (hex ? 0x80 : 0) | (alpha ? 0x100 : 0);
			i++;
		}
```
to support ASCII and the C1 mode of operation? Warrants a bunch of asserts against other cases so we don't silently do the wrong thing here, because it feels like it may have bad consequences.

Reply:
besides checking the value is greater than 0.. what are you thinking should be asserted?

Reply:
That dwInfoType == CT_CTYPE1, and that all characters are in the range [0, 256)